src/lib/Microsoft.Health.Events/EventHubProcessor/PartitionLockingBackgroundService.cs

```diff
@@ -0,0 +1,487 @@
+                var result = await leaseClient.AcquireAsync(TimeSpan.FromMinutes(1));
+                _ownedPartitions.Add(partitionToClaim);
+                _logger.LogTrace(result.ToString());
+            }
+        }
+
+        // registers a partition as a blob so that it can be leased later by a processor
+        private async Task RegisterPartitionBlobIfNotExists(BlobClient ownershipClient)
+        {
+            if (await ownershipClient.ExistsAsync())
+            {
+                return;
+            }
+            else
+            {
+                var eTag = default(ETag);
+                BlobUploadOptions ownershipClientOptions = new BlobUploadOptions()
+                {
+                    Conditions = new BlobRequestConditions()
+                    {
+                        IfMatch = eTag,
+                    },
+                };
+
+                try
+                {
+                    await ownershipClient.UploadAsync(BinaryData.FromString(string.Empty), ownershipClientOptions);
+                }
+                catch (RequestFailedException ex)
+                {
+                    if (ex.ErrorCode == BlobErrorCode.ConditionNotMet)
+                    {
+                        _logger.LogTrace($"Another processor uploaded a blob for partition id {ownershipClient.Name}");
+                        return;
+                    }
+
+                    throw;
+                }
+            }
+        }
+
+        private static Pageable<BlobItem> ListBlobs(BlobContainerClient blobContainerClient, string prefix)
+        {
+            try
+            {
+                var resultSegment = blobContainerClient.GetBlobs(BlobTraits.Metadata, BlobStates.None, prefix);
+                return resultSegment;
+            }
+            catch
+            {
+                throw;
+            }
+        }
+
+        private async Task<int> GetActiveProcessors(Pageable<BlobItem> resultSegment, BlobContainerClient containerClient, CancellationToken ct)
+        {
+            var activePods = new HashSet<string>();
+
+            foreach (BlobItem blobItem in resultSegment)
+            {
+                // if the processor has been updated in the last minute then consider it active
+                // otherwise delete it
+                if (blobItem.Metadata.TryGetValue("activeuntil", out var activeUntil))
+                {
+                    if (DateTime.Parse(activeUntil) > DateTime.UtcNow.AddMinutes(-1))
+                    {
+                        _logger.LogTrace($"Blob name: {blobItem.Name} is active");
+                        activePods.Add(blobItem.Name);
+                    }
+                    else
+                    {
+                        // todo: check the etag as a condition for delete
+                        _logger.LogTrace($"Deleting blob {blobItem.Name}. It was last active {activeUntil}.");
+                        await containerClient.DeleteBlobAsync(blobItem.Name);
```

Comment:
We should use E-Tag to delete and handle/ignore conflicts.

Reply:
This might be another case where we should elect a leader and only the leader can clean up old blobs.